### DISCUSSION TITLE
What kind of integrations do you want for xs-dev?

### DISCUSSION THREAD

**Discussion by HipsterBrown** [HipsterBrown/xs-dev#131] [Q&A]
At the moment, the only way to use xs-dev is through the CLI. What other ways do you want interact with it?

- VSCode extension?
- neovim plugin?
- menu bar / desktop app?
- docker image for CI/CD tasks
- node package with programmatic interface?

When looking at projects like DeviceScript, Espurino, PlatformIO, there's more ways to meet people where they are in terms of developer experience. Starting with breaking out the business logic into a "core" module for the CLI to consume, this should allow for folks to build their own integrations and tooling as well.

**Comment by bartmichu** [comment]
VS Code extension, definitely :)

**Reply by phoddie** [reply]
VS Code covers a lot of ground. There are so many ways to integrate there. What would your priorities be?

**Reply by bartmichu** [reply]
Here is what would make the whole experience feel better for me:
- initial setup (Install the Moddable SDK and build its tools, install the required development tools)
- updates notification and installation to everything mentioned above
- board/platform detection
- board connection status indicator
- tasks tailored to the board currently connected ([here is](https://github.com/bartmichu/moddable-extras/blob/master/vscode-tasks/tasks.json) what I'm using at the moment, not very smooth)
- clever links to your documentation

**Reply by phoddie** [reply]
Hey, @bartmichu, thanks for the thoughts. I like the goals. I'm not sure how to achieve it in all cases yet. Some quick thoughts. Maybe you or someone else has ideas on how to solve these.

> - initial setup (Install the Moddable SDK and build its tools, install the required development tools)
> - updates notification and installation to everything mentioned above

I think xs-dev takes care of most of this already. Integrating it into VS Code in a nice way would be mostly a UI project?

> - board/platform detection

That's more difficult. The `scan` feature of xs-dev does this and it is very useful. But, has limits because boards are not generally set-up for auto-detect. We know, for example, that is an ESP32 MCU but not which development board. It could be possible to make a guess at that by downloading some code to the device to probe further, but it will be imperfect (e.g. on many devices the display is write-only, so there's no way to query it, I2C bus scans on ESP32 are painfully slow, etc). Perhaps scan for MCU combined with selecting the board by name from the available ports for that MCU would be a good starting point?

> board connection status indicator

Like the above, this is sometimes possible but probably not in every case. 

> tasks tailored to the board currently connected ([here is](https://github.com/bartmichu/moddable-extras/blob/master/vscode-tasks/tasks.json) what I'm using at the moment, not very smooth)

Thanks for sharing that. Adding tasks to build & run on a give board is pretty straightforward. When you say it isn't "very smooth" is that because you have to create them manually or an other reason?

> clever links to your documentation

Expanding that, are you asking for inline documentation links when editing code?

**Reply by bartmichu** [reply]
> Perhaps scan for MCU combined with selecting the board by name from the available ports for that MCU would be a good starting point?

Definitely yes. If it remembers and completes from my past choices that would be perfect.

> When you say it isn't "very smooth" is that because you have to create them manually or an other reason?

It doesn't have any knowledge of what board is connected so tasks list is very visually polluted. Dedicated UI element (button) would be welcome as well.

> are you asking for inline documentation links when editing code?

OK I will try to describe it. I hope it is not too chaotic, I'm adding a picture to make it easier for me :)

So whenever I hover on something Moddable related (1,2,3,4), a pop-up appears with some basic documentation (like fig. A).
Clicking on (1) or (4) takes me to the source code in `$MODDABLE/`. Clicking on (2) opens up a browser with relevant documentation. While at imports, VSCode should also be aware of what can be imported and auto-complete paths.

I think a huge part of what I'm asking for is what VS Code does to JSDoc by default. Maybe adding JSDoc would give you this part of integration for free? Plus, it would benefit other environments as well.

![Screenshot](https://github.com/HipsterBrown/xs-dev/assets/4947154/00ee08cb-110e-4a1e-908f-f70d9d9728f5)

**Reply by phoddie** [reply]
Thanks for the detailed explanation and the (chaotic!) image. I like the ideas. I feel like we should try to stick close to what VS Code usually does. I'm not a heavy VS Code user, so it might be that what you describe is usual, but I'm not familiar with all of it. I'm happy to learn if you or someone else can point me in the right direction.

When working in TypeScript in VS Code, I get all sorts of useful information about Moddable SDK APIs from because of the [type declarations](https://github.com/Moddable-OpenSource/moddable/tree/public/typings) included in the Moddable SDK. Here's an example:

![image](https://github.com/HipsterBrown/xs-dev/assets/6856458/c0679289-a6c9-4c9a-8571-834b39d75b98)

If I add some JSDoc comments to timer.d.ts, then it will show some documentation when typing or hovering:

![image](https://github.com/HipsterBrown/xs-dev/assets/6856458/20ebecd9-9ca8-4228-a3d7-c84bc7430119)

That looks about the same as what VS Code shows for JavaScript built-ins like Promises. (In fact kind of like (A) in your image). The link gets you to the complete documentation (as your (2) suggests). That could also include a link to the sources.

FWIW – I'm pretty sure that the TypeScript declarations (and consequently JSDoc comments) can also be used when editing JavaScript with the correct configuration.

**Comment by DMGambone** [comment]
I'm late to the party here, but vscode debugging integration would absolutely be killer.  I like to run all my code in WSL2 on Windows, so being able to write the code in vscode and then run/debug in the same IDE with breakpoints and immediate feedback would be awesome.
